src/db/pool.cr

```diff
@@ -34,21 +38,49 @@ module DB
     @availability_channel : Channel(Nil)
     # signal how many existing connections are waited for
     @waiting_resource : Int32
+    # communicate when the pool is closed, to kill the Connection Reaper
+    @close_channel : Channel(Nil)
     # global pool mutex
     @mutex : Mutex
 
-    def initialize(@initial_pool_size = 1, @max_pool_size = 0, @max_idle_pool_size = 1, @checkout_timeout = 5.0,
-                   @retry_attempts = 1, @retry_delay = 0.2, &@factory : -> T)
+    def initialize(
+      @initial_pool_size = 1,
+      @max_pool_size = 0,
+      @max_idle_pool_size = 1,
+      @checkout_timeout = 5.0,
+      @retry_attempts = 1,
+      @retry_delay = 0.2,
+      @reaping_delay = 60.0,
+      @reaping_frequency = 60.0,
+      &@factory : -> T
+    )
       @availability_channel = Channel(Nil).new
+      @close_channel = Channel(Nil).new
       @waiting_resource = 0
       @inflight = 0
       @mutex = Mutex.new
 
       @initial_pool_size.times { build_resource }
     end
 
+    def start_reaper!
+      spawn(name: "connection_reaper") do
+        loop do
+          select
+          when @close_channel.receive?
+            break
+          when timeout @reaping_frequency.seconds
+            sync do
+              @idle.each_with_index { |conn, i| reap_connection! conn, i }
+            end
```

Comment:
I don't think syncing the entire reaper sequence is going to work. `sync` acquires a lock at the entry and only releases it on exit. So while the block is running, all other sync operations are blocked (they'll be unblocked at `unsync`, but that's only a minor relief). Most of the time in the reaper loop is spent sleeping and there is definitely no need to sync that.
Syncs should cover each `reap_connection!` individually and without the sleep.

Reply:
👍 that makes sense -- I reverted the commit, but then added a bit to a test that will help illustrate the problem I was looking to fix with the current implementation. 

Let me know if you have any suggestions, but the gist of the issue is that the sleep isn't actually doing anything in terms of enforcing the delay to prevent the `@idle` pool from being exhausted while waiting on health checks.

Reply:
https://github.com/crystal-lang/crystal-db/pull/170/commits/56e0a35c18f32f8d1e360504976baad2b191f151 shows the issue I was trying to get around, but I can see that my approach isn't correct for that.

I added some puts debugging to track which "attempt" the reaper is on when things are being checked, which helps me see where some of the timing-dependent tests were. 

LMK if you have any immediate thoughts, but I may need to rethink my approach a bit here -- my only real goal is to prevent having all the `@idle` connections be removed at the same time. 

I'm going to give a shot at moving things into a `@health_check` pool (or possibly just using a buffered channel?), but I'm open to any other ideas to help simplify the code and make the testing story a bit more reliable.

Reply:
Definitely a bit of a 🤦 moment.

Code is much simpler now, and I *think* the tests should be much less brittle, or at the very least, much easier to pad to help slower CI boxes compensate now.